chainer/training/extensions/parameter_statistics.py

```diff
@@ -0,0 +1,233 @@
+import numpy
+
+import cupy
+
+from chainer import cuda
+from chainer import reporter
+from chainer import training
+from chainer.training import extension
+
+
+def _iterable(x):
+    if isinstance(x, (list, tuple)):
+        return x
+    return x,
+
+
+def _prefix_statistics(prefix, stats):
+    """Prefix all keys in a statistic dictionary."""
+    for key in list(stats.keys()):
+        stats['{}/{}'.format(prefix, key)] = stats.pop(key)
+    return stats
+
+
+def _statistic_key(link, param_names, attr_names):
+    """Generate a statistic dictionary key based on context."""
+    param_names = _iterable(param_names)
+    attr_names = _iterable(attr_names)
+
+    link_name = 'None' if not hasattr(link, 'name') else link.name
+    param_name = '-'.join(param_names)
+    attr_name = '-'.join(attr_names)
+
+    return '{}/{}/{}'.format(link_name, param_name, attr_name)
+
+
+def _flatten_link(link, param_names, attr_names):
+    """Flatten a link into an array."""
+    param_names = _iterable(param_names)
+    attr_names = _iterable(attr_names)
+
+    params = []
+    for param in link.params():
+        if param.name in param_names:
+            for attr_name in attr_names:
+                p = getattr(param, attr_name)
```

Comment:
At the moment, we assume that the user is aware of the internal names of the links, e.g. `W.data` or `gamma.grad`. Those names are also expected as args in the updated interface (`target=(('W','data'),('W','grad'))`).

It is flexible, but might not be according to the design principles of Chainer. How do you think we should let the user specify which parts of the links to monitor?